### PR TITLE
Unit tests.

### PR DESCRIPTION
**Pull request by alexkart**
Edit phpunit config to allow code coverage.
Added unit tests for the uncovered code.

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  |

### CODE REVIEW

**Comment by samdark** [comment]
@alexkart please resolve conflicts. Thanks.

File: src/Client.php [hunk 1]
```diff
@@ -407,7 +407,7 @@ public function afterSend($request, $response)
      * @return Request request instance.
      * @throws \yii\base\InvalidConfigException
      */
-    private function createRequestShortcut($method, $url, $data, $headers, $options)
+    protected function createRequestShortcut($method, $url, $data, $headers, $options)
```
Comment (samdark) [inline src/Client.php]: That is considered backwards compatibility break and should not be done.

File: src/Response.php [hunk 2]
```diff
@@ -196,9 +196,6 @@ private function normalizeCookieParamName($rawName)
     private function getParser()
     {
         $format = $this->getFormat();
-        if ($format === null) {
```
Comment (samdark) [inline src/Response.php]: Why?

File: tests/ClientTest.php [hunk 3]
```diff
@@ -34,6 +34,11 @@ public function testSetupFormatters()
         $formatter = $client->getFormatter('testConfig');
         $this->assertTrue($formatter instanceof UrlEncodedFormatter);
         $this->assertEquals(PHP_QUERY_RFC3986, $formatter->encodingType);
+
+        $unrecognizedFormat = 'unrecognizedFormat';
```
Comment (samdark) [inline tests/ClientTest.php]: It is better to move it to separate test-method such as `testSetupUnrecognizedFormat`.

File: tests/ClientTest.php [hunk 4]
```diff
@@ -93,6 +98,11 @@ public function testSetupParsers()
 
         $parser = $client->getParser('testConfig');
         $this->assertTrue($parser instanceof UrlEncodedParser);
+
+        $unrecognizedParser = 'unrecognizedParser';
```
Comment (samdark) [inline tests/ClientTest.php]: It is better to move it to separate test-method such as `testSetupUnrecognizedParser`.

File: tests/ClientTest.php [hunk 5]
```diff
@@ -204,4 +214,55 @@ public function testCreateResponseWithHeadersEqualToEmptyArray()
...
+    {
+        $method = 'POST';
+        $url = 'url';
+        $data = ['data'];
+        $headers = ['headers'];
+        $options = ['options'];
+
+        $client = new Client();
+        /** @var Request $request */
+        $request = $this->invoke($client, 'createRequestShortcut', [$method, $url, $data, $headers, $options]);
```
Comment (samdark) [inline tests/ClientTest.php]: No need to do if we'll leave it as public...

File: tests/MessageTest.php [hunk 6]
```diff
@@ -175,5 +175,18 @@ public function testSetupData()
         ];
         $message->addData($additionalData);
         $this->assertEquals(array_merge($data, $additionalData), $message->getData());
+
+        $this->expectException('\yii\base\Exception');
```
Comment (samdark) [inline tests/MessageTest.php]: Better to extract it into separate test-method.

File: tests/ResponseTest.php [hunk 7]
```diff
@@ -140,6 +140,11 @@ public function testGetStatusCode()
             $statusCode,
         ]]);
         $this->assertEquals($statusCode, $response->getStatusCode());
+
+        $this->expectException('\yii\httpclient\Exception');
```
Comment (samdark) [inline tests/ResponseTest.php]: Better to extract it into separate test-method.

File: tests/XmlFormatterTest.php [hunk 8]
```diff
@@ -33,6 +34,14 @@ public function testFormat()
 XML;
         $this->assertEqualsWithoutLE($expectedContent, $request->getContent());
         $this->assertEquals('application/xml; charset=UTF-8', $request->getHeaders()->get('Content-Type'));
+
+        $request->setData('data');
```
Comment (samdark) [inline tests/XmlFormatterTest.php]: Better to extract it into separate test-method.

File: src/Response.php (hunk 2 above)
Comment (alexkart) [reply src/Response.php]: getFormat() never returns null, code inside if statement can never be reached

File: tests/ClientTest.php (hunk 5 above)
Comment (alexkart) [reply tests/ClientTest.php]: You mean private?

Comment (samdark) [reply tests/ClientTest.php]: https://github.com/yiisoft/yii2-httpclient/pull/187#discussion_r279402801

Comment (alexkart) [reply tests/ClientTest.php]: I'm not sure I understand what you mean. `createRequestShortcut` was private and I changed it to protected to be able to invoke it in the test. Why is that considered backward compatibility break?

Comment (samdark) [reply tests/ClientTest.php]: I misread it yesterday. It's alright. Sorry.

**Comment by samdark** [comment]
Merged. Thank you very much!

**Pull request merged**